worlds/osrs/__init__.py

```diff
@@ -0,0 +1,616 @@
+                state.has(ItemNames.QP_Below_Ice_Mountain, self.player)
+            return
+        if region_row.name == "Dwarven Mountain Pass" and outbound_region_name == "Anvil*":
+            entrance.access_rule = lambda state: state.has(ItemNames.QP_Dorics_Quest, self.player)
+            return
+
+        print(f"Special rules required to access region {outbound_region_name} from {region_row.name}")
+
+    def roll_locations(self):
+        locations_required = 0
+        for item_row in self.item_rows:
+            locations_required += item_row.count
+
+        locations_added = 1 # At this point we've already added the starting area, so we start at 1 instead of 0
+
+        # Quests are always added
+        for i in range(len(self.location_rows)):
+            location_row = self.location_rows[i]
+            if location_row.category in ["Quest", "Points", "Goal"]:
+                self.create_and_add_location(i)
+                if location_row.category == "Quest":
+                    locations_added += 1
+
+        # Build up the weighted Task Pool
+        rnd = self.multiworld.per_slot_randoms[self.player]
+
+        # Start with the minimum general tasks
+        general_tasks = [task for task in self.locations_by_category["General"]]
+        if not self.multiworld.progressive_tasks[self.player]:
+            rnd.shuffle(general_tasks)
+        else:
+            general_tasks.reverse()
+        for i in range(0, self.multiworld.minimum_general_tasks[self.player]):
+            task = general_tasks.pop()
+            self.add_location(task)
+            locations_added += 1
+
+        general_weight = self.multiworld.general_task_weight[self.player] if len(general_tasks) > 0 else 0
+        combat_tasks = [task for task in self.locations_by_category["Combat"]
+                        if task.skills[0].level <= int(self.multiworld.max_combat_level[self.player])]
+        #if not self.multiworld.progressive_tasks[self.player]:
+        rnd.shuffle(combat_tasks)
+        combat_tasks = combat_tasks[0:self.multiworld.max_combat_tasks[self.player]]
+        combat_weight = self.multiworld.combat_task_weight[self.player] if len(combat_tasks) > 0 else 0
+
+        prayer_tasks = [task for task in self.locations_by_category["Prayer"]
+                        if task.skills[0].level <= int(self.multiworld.max_prayer_level[self.player])]
+        if not self.multiworld.progressive_tasks[self.player]:
+            rnd.shuffle(prayer_tasks)
+        else:
+            prayer_tasks.reverse()
+        prayer_tasks = prayer_tasks[0:self.multiworld.max_prayer_tasks[self.player]]
+        prayer_weight = self.multiworld.prayer_task_weight[self.player] if len(prayer_tasks) > 0 else 0
+
+        magic_tasks = [task for task in self.locations_by_category["Magic"]
+                       if task.skills[0].level <= int(self.multiworld.max_magic_level[self.player])]
```

Comment:
just a suggestion, but you could *probably* automate this with a `getattr` hack instead of duplicating the same code for every task type.

```
tasks_per_task_type: Dict[str, List[LocationRow]] = {}
weights_per_task_type: Dict[str, int] = {}

task_types = ["prayer", "magic", "runecraft", ...]
for task_type in task_types:
    max_level_for_task_type = getattr(self.options, f"max_{task_type}_level")
    max_amount_for_task_type = getattr(self.options, f"max_{task_type}_tasks")
    tasks_for_this_type = [task for task in self.locations_by_category[task_type]
                                       if task.skills[0].level <= max_level_for_task_type]
    if not self.multiworld.progressive_tasks[self.player]:
        rnd.shuffle(tasks_for_this_type)
    else:
        tasks_for_this_type.reverse()

    tasks_for_this_type = tasks_for_this_type[:max_amount_for_task_type]

    tasks_per_task_type[task_type] = tasks_for_this_type
    weights_per_task_type[task_type] = getattr(self.options, f"{task_type}_task_weight") if tasks_for_this_type else 0

...
#  use the dicts instead of making hand-written lists
...
```

You can decide yourself how evil you think this is.

Also included: `some_list[:end] instead of some_list[0:end]` as well as `if some_list` instead of `if len(some_list) > 0`